### ISSUE TITLE
Fails to inject SecurityContext into Helloworld-CDI2-SE example

### ISSUE THREAD

**Issue by mkarg** [eclipse-ee4j/jersey#4208]
Given `SecurityContext` is injected into `examples/helloworld-cdi2-se`'s `HelloWorldResource`
```java
    public String getHello(@PathParam("name") String name, @Context SecurityContext securityContext) {
        try {
            securityContext.getUserPrincipal();
        } catch (final Exception e) {
            e.printStackTrace();
            throw e;
        }
```
When requesting the resource
```bash
curl http://localhost:8080/helloworld/Some%20Name
```
Then Jersey fails with exception
```
java.lang.ClassCastException: org.glassfish.jersey.internal.util.collection.Refs$DefaultRefImpl cannot be cast to org.glassfish.jersey.server.ContainerRequest
        at org.glassfish.jersey.server.ContainerRequest$Proxy$_$$_WeldClientProxy.getSecurityContext(Unknown Source)
        at org.glassfish.jersey.server.internal.process.SecurityContextInjectee.getUserPrincipal(SecurityContextInjectee.java:62)
        at org.glassfish.jersey.examples.helloworld.cdi2se.HelloWorldResource.getHello(HelloWorldResource.java:40)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
        at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:219)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80)
        at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:232)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680)
        at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:353)
        at org.glassfish.grizzly.http.server.HttpHandler$1.run(HttpHandler.java:200)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:569)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:549)
        at java.lang.Thread.run(Thread.java:748)

```

**Comment by mkarg** [comment]
@senivam Any idea how to fix that?

**Comment by mkarg** [comment]
@jansupol Kindly requesting a fix for 2.29.1 as per your offer in the mailing list. :-)

**Comment by mkarg** [comment]
Would be cool if Team Jersey finds the time to at least confirm that this is a bug in Jersey (not in Weld). :-)

FYI, when adding Jersey's Role Allowed Feature to the example, the roles check works great, so it seems to happen _later_ in the process chain (because that feature *can* access the provided security context).

**Comment by pa314159** [comment]
I think the problem is the same as for issue 3753; I tested with my PR and the CCE is gone:

https://github.com/eclipse-ee4j/jersey/pull/4219

**Comment by mkarg** [comment]
@pa314159 That sounds great! I hope Team Jersey picks up your PR soon. :-)

**Comment by mkarg** [comment]
I confirm this problem is solved by #4219. Thank you, @pa314159! 👏

**Comment by pa314159** [comment]
Unfortunately I found another issue with injection by @Context in InjectingConstraintValidatorFactory. :(

I'll try to figure out what is going on in cdi2-se.

**Comment by pa314159** [comment]
I implemented an workaround for the injection issue mentioned few adys ago, but as @mkarg said, that proves the is still a bug in cdi2-se.

However the CCE is now solved by https://github.com/eclipse-ee4j/jersey/pull/4219

**Comment by mkarg** [comment]
Just to make it pretty clear: Will solving the CCE *produce* that new bug, or is that new bug *unrelated* of the solution of the CCE? Because in the latter case, we could simply put the solution for the CCE in a standalone PR and merge that one at least.

**Comment by pa314159** [comment]
There is no _new_ bug, it's your comment which I agree with: https://github.com/eclipse-ee4j/jersey/pull/4219#commitcomment-34643045

Both https://github.com/eclipse-ee4j/jersey/issues/4208 and https://github.com/eclipse-ee4j/jersey/issues/3753 were caused by SuplierBeanBridge that returned the same value for Bean.getId(). These are now solved.

**Comment by mkarg** [comment]
But with https://github.com/eclipse-ee4j/jersey/pull/4219/commits/0345aef488d144c40fcbd9a88555d3f9a004f95d you did not fix a bug with `@Context`, you simply did not use `@Context` at all. @senivam Can you please judge about that?

**Comment by pa314159** [comment]
No, I didn't say I fixed that :) it's just an workaround to a possible different bug which was not the subject of these two issues.

What I fixed is the CCE reported here and the ILA reported by #3753.

Without the implementation of SupplierBeanBridge.getBeanClass(), any invocations of SupplierBeanBridge.getId() would have returned something like **java.lang.Object#jersey**. 

Instances of SupplierBeanBridge  are put in a map using getId() as a key and obviously any new insertion overrides the previous one.
